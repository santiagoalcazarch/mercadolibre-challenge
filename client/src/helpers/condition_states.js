
export const itemCondition = {
  new: 'Nuevo',
  used: 'Usado',
}