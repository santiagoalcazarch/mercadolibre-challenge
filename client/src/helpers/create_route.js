

/**
 * Retorna una cadena de texto con el fin de optimizar la
 * creaciÃ³n de query params en una ruta
 * @param {string} route 
 * @param {object || array} params Object or array
 * @returns Cadena de texto con la ruta y los parametros
 */
export const getRouteWithQueryParams = ( route, params ) => {

  const queryParams = new URLSearchParams(params).toString();

  return route + "?" + queryParams;
}