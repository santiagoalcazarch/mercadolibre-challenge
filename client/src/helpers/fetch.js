
import { BASE_URL } from "../config/api/api_routes";

/**
 * Ejecuta y retorna la función [fetch] para realizar un llamado HTTP(s)
 * @param {string} route - Ruta a la que será dirigído el Fetch
 * @param {string} method - Metodo del fetch
 * @param {object} headers - Header del fetch
 * @param {any} others - Cualquier otra propiedad para agregar al fetch 
 * @returns 
 */
function excecuteFetch({ route, method, headers, ...others }) {
  const config = {
    method,
    headers: headers ? headers : new Headers(),
    ...others
  };
  const request = new Request(BASE_URL + route, config);
  return fetch(request);
}
  
export default excecuteFetch;
