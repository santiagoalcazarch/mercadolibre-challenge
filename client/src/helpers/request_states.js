
export const requestState = {
  SUCCESS: 'success',
  FAILED: 'failed',
  INIT: 'init',
  LOADING: 'loading'
}