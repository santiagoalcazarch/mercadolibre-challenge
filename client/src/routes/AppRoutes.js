import React from 'react'
import { BrowserRouter, Route, Routes } from 'react-router-dom'
import ItemPage from '../ecommerce/pages/item/ItemPage'
import Layout from '../ecommerce/pages/home/Layout'
import ListItemPage from '../ecommerce/pages/list-item/ItemListPage'
import Welcome from '../ecommerce/pages/welcome/welcome'
import { HOME, ITEM, LIST_ITEM } from './routes'

const AppRoutes = () => {
  return (
    <BrowserRouter>
      <Routes>
        <Route path={HOME} element={<Layout />} >
          <Route path={HOME} element={ <Welcome /> } />
          <Route path={LIST_ITEM} element={<ListItemPage />} />
          <Route path={ITEM} element={<ItemPage />} />
        </Route>
      </Routes>
    </BrowserRouter>
  )
}

export default AppRoutes