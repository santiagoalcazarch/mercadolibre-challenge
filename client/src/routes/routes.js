

export const HOME = "/";

export const LIST_ITEM = "/items";

export const ITEM = "/items/:id";
export const ITEM_FUNC = (itemId) => `/items/${itemId}`;