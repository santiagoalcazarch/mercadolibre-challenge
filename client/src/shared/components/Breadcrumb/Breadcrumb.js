
import React from 'react'
import PropTypes from 'prop-types'

const Breadcrumb = ({ categories }) => {
  return (
    <div className="container p-0">
      <div className="breadcrumb-container">
        
        <ul className="breadcrumb-items p-0">
          {
            categories.map( (category, i) => 
              <li className="breadcrumb-item p-0">
                <a href="#">
                  <span> { category } </span>
                </a>
                {
                  i < categories.length - 1 && 
                  <span className="mx-1"> {'>'} </span>
                }
              </li>
            )
          }
        </ul>
      </div>
    </div>
  )
}

Breadcrumb.propTypes = {
  categories: PropTypes.array.isRequired,
}

export default Breadcrumb