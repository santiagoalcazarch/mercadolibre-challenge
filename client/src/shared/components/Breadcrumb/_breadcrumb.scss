
.breadcrumb-container{
  padding: 16px 0;

  ul.breadcrumb-items{
    display: inline;

    li.breadcrumb-item{
      display: inline;

      color: $mainGray;
      
      a:link, a:visited, a:hover, a:focus, a:active{
        color: $mainGray;
        text-decoration: none;
        transition: color 0.2s ease;
      }
      
      a:hover{
        color: $mainBlack;
      }
    }

    li.breadcrumb-item::before{
      content: "";
      display: none;
    }
    
  }
}