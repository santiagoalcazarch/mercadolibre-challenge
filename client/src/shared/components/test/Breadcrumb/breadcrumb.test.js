import { mount } from "enzyme"
import { MemoryRouter } from "react-router-dom";
import { LIST_ITEM } from "../../../../routes/routes";
import Breadcrumb from "../../Breadcrumb/Breadcrumb";
import { requestState } from "../../../../helpers/request_states";

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from "react-redux";

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const storeInitState = {
  items: {
    itemsList: { categories: [], items: [] },
    itemInfo: { categories: [], item: {} },
    requestState: requestState.INIT
  }
}

let store = mockStore(storeInitState);

describe('Breadcrumb test', () => {

  beforeEach( () => {
    store = mockStore(storeInitState);
  })

  test('Debería hacer match con el snapshot', () => {
    const wrapper = mount( 
      <Provider store={store}>
        <MemoryRouter initialEntries={[`${LIST_ITEM}`]}> 
          <Breadcrumb />
        </MemoryRouter>
      </Provider>
    )
    expect(wrapper).toMatchSnapshot();
  });

  test('Debería renderizar las categorias - ItemList', async () => {

    let cateogriesList = ["any", "category"];
    store = mockStore({
      items: {
        itemsList: { categories: cateogriesList, items: [] },
        itemInfo: { categories: [], item: {} },
        requestState: requestState.INIT
      }
    });
    
    const wrapper = mount( 
      <Provider store={store}>
        <MemoryRouter initialEntries={["/items?search=impresora"]}> 
          <Breadcrumb />
        </MemoryRouter>
      </Provider>
    )

    const categoriesRenderTest = wrapper.find(".breadcrumb-item").map( 
      ( category ) => category.find("a span").text().trim()
    )

    expect(categoriesRenderTest).toEqual(cateogriesList);
  });

  test('Debería renderizar las categorias - Item', async () => {

    let cateogriesList = ["any", "category"];
    store = mockStore({
      items: {
        itemsList: { categories: [], items: [] },
        itemInfo: { categories: cateogriesList, item: {} },
        requestState: requestState.INIT
      }
    });
    
    const wrapper = mount( 
      <Provider store={store}>
        <MemoryRouter initialEntries={["/items/MCO432343"]}> 
          <Breadcrumb />
        </MemoryRouter>
      </Provider>
    )

    const categoriesRenderTest = wrapper.find(".breadcrumb-item").map( 
      ( category ) => category.find("a span").text().trim()
    )

    expect(categoriesRenderTest).toEqual(cateogriesList);
  });

});
