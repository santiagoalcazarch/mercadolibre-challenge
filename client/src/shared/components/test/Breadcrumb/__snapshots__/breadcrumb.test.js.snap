// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb test Deber√≠a hacer match con el snapshot 1`] = `
<div
  className="container p-0"
>
  <div
    className="breadcrumb-container"
  >
    <ul
      className="breadcrumb-items p-0"
    />
  </div>
</div>
`;
